test/integration/resolve_uris.cc

```diff
@@ -418,3 +419,67 @@ TEST(ResolveURIs, BadCallback)
   EXPECT_EQ(kDiffuseUri, diffuseElemUri);
   EXPECT_EQ(kNormalUri, normalElemUri);
 }
+
+/////////////////////////////////////////////////
+TEST(ResolveURIs, ResolvedRelativeURIs)
+{
+  const std::string testFile =
+    sdf::testing::TestFile("sdf", "resolve_relative_uris.sdf");
+  const std::string sdfDir = sdf::testing::TestFile("sdf");
+
+  sdf::ParserConfig config;
+  config.SetStoreResovledURIs(true);
+  EXPECT_TRUE(config.StoreResolvedURIs());
+
+  size_t callbackCount = 0;
+  config.SetFindCallback(
+      [=, &callbackCount](const std::string &)
+      {
+        callbackCount++;
+        return "";
+      });
+
+  sdf::Root root;
+  auto errors = root.Load(testFile, config);
+  EXPECT_TRUE(errors.empty());
+  // no user callbacks should be invoked
+  EXPECT_EQ(0, callbackCount);
+
+  auto world = root.WorldByIndex(0);
+  ASSERT_NE(nullptr, world);
+
+  /// Skybox Cubemap Uri
+  auto sky = world->Scene()->Sky();
+  auto cubemapUri = sky->CubemapUri();
+
+  /// Visual mesh and material script uri
+  auto model = world->ModelByName("model");
+  ASSERT_NE(nullptr, model);
+  auto link = model->LinkByName("link");
+  ASSERT_NE(nullptr, link);
+  auto visual = link->VisualByName("visual");
+  ASSERT_NE(nullptr, visual);
+  auto meshVisualUri = visual->Geom()->MeshShape()->Uri();
+  auto scriptUri = visual->Material()->ScriptUri();
+
+  // Collision mesh uri
+  auto meshCol = link->CollisionByName("collision");
+  ASSERT_NE(nullptr, meshCol);
+  auto meshColUri = meshCol->Geom()->MeshShape()->Uri();
+
+  EXPECT_NE(std::string::npos, cubemapUri.find("dummy_cubemap.dds"));
+  EXPECT_NE(std::string::npos, scriptUri.find("dummy_material.material"));
+  EXPECT_NE(std::string::npos, meshVisualUri.find("dummy_mesh.dae"));
+  EXPECT_NE(std::string::npos, meshColUri.find("dummy_mesh.dae"));
+  EXPECT_EQ(meshVisualUri, meshColUri);
+
+  EXPECT_TRUE(std::filesystem::path(cubemapUri).is_absolute());
+  EXPECT_TRUE(std::filesystem::path(scriptUri).is_absolute());
+  EXPECT_TRUE(std::filesystem::path(meshVisualUri).is_absolute());
+  EXPECT_TRUE(std::filesystem::path(meshColUri).is_absolute());
+
+  EXPECT_TRUE(sdf::filesystem::exists(cubemapUri));
+  EXPECT_TRUE(sdf::filesystem::exists(scriptUri));
+  EXPECT_TRUE(sdf::filesystem::exists(meshVisualUri));
+  EXPECT_TRUE(sdf::filesystem::exists(meshColUri));
+}
```

Comment:
I would like to see a test where the input is an SDF string, not a file, so as to exercise the situation where `sdf::Element::FilePath` returns an empty string. It's not clear to me what `parent_path` returns in that case.

Reply:
added test for loading sdf from string. In this case, `parent_path` will be empty string so it's then up to the user callback function to search for the uri file. [c130219](https://github.com/gazebosim/sdformat/pull/1373/commits/c130219fce8a874cca1e419fefa4d524df9e16b8)